### ISSUE TITLE
Nanboxing seems to break Eigen::SparseMatrix construction

### ISSUE THREAD

**Issue by chenyang-tri** [RobotLocomotion/drake#18218]
When constructing an `Eigen::SparseMatrix<drake::symbolic::Expression>` from triplets, non-numeric `drake::symbolic::Expression`s are treated as zero. See test below to reproduce this bug. 

This bug occurs only after and including v1.6.0 of Drake, and not on v1.5.0, so I'm suspecting it has to do with the changes introduced here: #17544 

```cpp
#include <gtest/gtest.h>
#include "drake/common/symbolic/expression.h"

namespace {
  using std::vector;
  using drake::symbolic::Variable;
  using drake::symbolic::Expression;

  typedef Eigen::Triplet<Expression> TE;
  typedef Eigen::SparseMatrix<Expression> SE;
  typedef Eigen::Matrix<Expression, 5, 5> ME;

  GTEST_TEST(SymbolicMatricesTest, SparseMatrices) {
    std::vector<TE> tripletList;
    tripletList.reserve(10);
    tripletList.push_back(TE(1,1,1.1));
    tripletList.push_back(TE(2,2,Variable("x")*2.));
    SE M(5,5);
    M.setFromTriplets(tripletList.begin(), tripletList.end());
    EXPECT_EQ(M.coeff(0,0), 0.);
    EXPECT_EQ(M.coeff(1,1), 1.1);
    EXPECT_EQ(M.coeff(2,2).to_string(), (Variable("x")*2.).to_string());
  }

  GTEST_TEST(SymbolicMatricesTest, DenseMatrices) {
    ME M;
    M(0,0) = 0.;
    M(1,1) = 1.1;
    M(2,2) = Variable("x")*2.;
    EXPECT_EQ(M(0,0), 0.);
    EXPECT_EQ(M(1,1), 1.1);
    EXPECT_EQ(M(2,2).to_string(), (Variable("x")*2.).to_string());
  }

}
```

**Comment by jwnimmer-tri** [comment]
I've posted a slight variation of the test case in #18219.  Still investigating.

**Comment by jwnimmer-tri** [comment]
Thanks for the report and nice test case.  I've posted the fix in #18219.
